File: lib/hyprdesk/grid.py
"""Phone-style placement grid for the desktop card fleet.

Each monitor divides into grid_cols × grid_rows cells (widgets.conf,
defaults 12×6) inside a fixed outer margin. A widget occupies its
top-left cell plus an auto span of ceil(size/cell); cell coordinates are
resolution-proof and make overlap a set-intersection check instead of
pixel math. The bottom strip (pet walkway + dock) stays out of the grid
via BOTTOM_RESERVE.

COORDINATE SPACE: every function takes WORKAREA dimensions (monitor size
minus Hyprland's reserved insets — waybar reserves 30px top here) and
returns WORKAREA-relative pixels. Layer-shell margins are workarea-
relative, so grid origins map 1:1 onto gtk_layer_set_margin values; a
caller working in absolute monitor pixels (hyprctl layers) must subtract
the reserved offsets first.
"""
import math

from .theme import conf

MARGIN = 24          # outer gutter, px (matches the fleet's classic offsets)
BOTTOM_RESERVE = 68  # pet strip + dock clearance at the screen bottom


def geometry(mon_w, mon_h, c=None):
    """(cols, rows, cell_w, cell_h) for a monitor, from widgets.conf."""
    c = conf() if c is None else c
    try:
        cols = max(2, min(24, int(c.get("grid_cols", 12))))
        rows = max(2, min(16, int(c.get("grid_rows", 6))))
    except (TypeError, ValueError):
        cols, rows = 12, 6
    cell_w = (mon_w - 2 * MARGIN) / cols
    cell_h = (mon_h - MARGIN - BOTTOM_RESERVE) / rows
    return cols, rows, cell_w, cell_h


def origin(mon_w, mon_h, col, row, c=None):
    """Top-left pixel of a cell (monitor-local)."""
    cols, rows, cw, ch = geometry(mon_w, mon_h, c)
    col = max(0, min(cols - 1, int(col)))
    row = max(0, min(rows - 1, int(row)))
    return round(MARGIN + col * cw), round(MARGIN + row * ch)


def cell_at(mon_w, mon_h, x, y, c=None):
    """Nearest cell (col, row) for a monitor-local pixel position."""
    cols, rows, cw, ch = geometry(mon_w, mon_h, c)
    col = round((x - MARGIN) / cw)
    row = round((y - MARGIN) / ch)
    return max(0, min(cols - 1, col)), max(0, min(rows - 1, row))


def span(mon_w, mon_h, w, h, c=None):
    """(col_span, row_span) cells a w×h px widget occupies."""
    _, _, cw, ch = geometry(mon_w, mon_h, c)
    return max(1, math.ceil(w / cw)), max(1, math.ceil(h / ch))


def clamp(mon_w, mon_h, col, row, col_span, row_span, c=None):
    """Clamp a cell position so the whole span stays on the grid."""
    cols, rows, _, _ = geometry(mon_w, mon_h, c)
    return (max(0, min(cols - col_span, int(col))),
            max(0, min(rows - row_span, int(row))))


def cells_of(col, row, col_span, row_span):
    """The set of (col, row) cells covered — overlap = set intersection."""
    return {(col + i, row + j)
            for i in range(col_span) for j in range(row_span)}

File: rust/hyprdesk/src/grid.rs
//! The phone-style placement grid, ported from lib/hyprdesk/grid.py.
//!
//! THE HIGHEST-STAKES PARITY IN THE MIGRATION: hypr-arrange (python,
//! staying python) WRITES `<id>_col/_row` cells that a rust cardhost
//! READS. The two implementations compute pixels from those cells
//! independently — one rounding difference and every card jumps on the
//! next reload. `examples/grid_parity.rs` sweeps both against each other;
//! run it whenever either side changes.
//!
//! COORDINATE SPACE: workarea dimensions in, workarea-relative pixels out
//! (layer margins are workarea-relative). Python's round() is banker's
//! rounding — round-half-to-even — and Rust's f64::round() is
//! round-half-away-from-zero. They disagree on exact .5 values, which
//! cell_w fractions can and do produce. `pyround` below is the python
//! semantics, verified by the parity sweep.

pub const MARGIN: f64 = 24.0;
pub const BOTTOM_RESERVE: f64 = 68.0;

/// python round(): banker's rounding to the nearest even integer.
fn pyround(x: f64) -> i32 {
    let floor = x.floor();
    let frac = x - floor;
    if (frac - 0.5).abs() < 1e-9 {
        // exactly half: to even
        let f = floor as i64;
        (if f % 2 == 0 { f } else { f + 1 }) as i32
    } else {
        x.round() as i32
    }
}

pub struct Grid {
    pub cols: i32,
    pub rows: i32,
    pub cell_w: f64,
    pub cell_h: f64,
}

pub fn geometry(mon_w: f64, mon_h: f64) -> Grid {
    let cols = crate::conf_get("grid_cols", "12")
        .parse::<i32>()
        .unwrap_or(12)
        .clamp(2, 24);
    let rows = crate::conf_get("grid_rows", "6")
        .parse::<i32>()
        .unwrap_or(6)
        .clamp(2, 16);
    Grid {
        cols,
        rows,
        cell_w: (mon_w - 2.0 * MARGIN) / cols as f64,
        cell_h: (mon_h - MARGIN - BOTTOM_RESERVE) / rows as f64,
    }
}

pub fn origin(mon_w: f64, mon_h: f64, col: i32, row: i32) -> (i32, i32) {
    let g = geometry(mon_w, mon_h);
    let col = col.clamp(0, g.cols - 1);
    let row = row.clamp(0, g.rows - 1);
    (
        pyround(MARGIN + col as f64 * g.cell_w),
        pyround(MARGIN + row as f64 * g.cell_h),
    )
}

pub fn cell_at(mon_w: f64, mon_h: f64, x: f64, y: f64) -> (i32, i32) {
    let g = geometry(mon_w, mon_h);
    let col = pyround((x - MARGIN) / g.cell_w);
    let row = pyround((y - MARGIN) / g.cell_h);
    (col.clamp(0, g.cols - 1), row.clamp(0, g.rows - 1))
}

pub fn span(mon_w: f64, mon_h: f64, w: f64, h: f64) -> (i32, i32) {
    let g = geometry(mon_w, mon_h);
    (
        ((w / g.cell_w).ceil() as i32).max(1),
        ((h / g.cell_h).ceil() as i32).max(1),
    )
}

pub fn clamp(mon_w: f64, mon_h: f64, col: i32, row: i32, cs: i32, rs: i32) -> (i32, i32) {
    let g = geometry(mon_w, mon_h);
    (
        col.clamp(0, (g.cols - cs).max(0)),
        row.clamp(0, (g.rows - rs).max(0)),
    )
}
